banshee-nomad/banshee-bn-backend
# #168563609 Notifications for new travel request(in-app)

#### What does this PR do?
Have an admin user receive an in-app notification when a new trip request is initiated
#### Description of Task to be completed?
- [x] Install socket.io and sockect.io-client
- [x] Create an in-app client
- [x] Create socket io server
- [x] Create endpoint that test in-app feature from through a browser
- [x] Modify new trip request method
- [x] Create a function that setup connection for a namespace and emits data to
  specific recipients in a room(company)

#### How should this be manually tested?
1. Run `npm run devstart`
 2. Launch postman
 3. Make a post request to `/api/v1/auth/login`  with valid credentials
 4. Copy the token in the response body and paste it in notepad.
 5. Open a browser and enter the URL `http://localhost:8000/api/v1/trips/test/in_app`. You see the following icon below.
![notifyIco](https://user-images.githubusercontent.com/28379439/66023134-2bce5980-e4e8-11e9-820e-6007f2b8b4ab.JPG)

 6. Make a post request to `/api/v1/trips`.Use the authorization(token) in the request header. Provide valid data in the request body. See the example below

```
{

    "type": "oneway",
    "from": "3dd3b34a-7554-425e-xxxx-36afda199614",
    "departureDate": "2020-09-09",
    "reason": "Just wanna travel",
    "destinations": [
    {
      "to": "3dd3b34a-7554-455e-a688-xxxxxxxxxx4",
      "accomodation": "3dd3b34a-7554-425e-axxxxxxxxxxxxx9"
    }]
  }

```

  7. Hit send
#### What are the relevant pivotal tracker stories?
[#168563609](https://www.pivotaltracker.com/story/show/168563609)
#### Any background context you want to add?
NIL
#### Screenshots
![inApp](https://user-images.githubusercontent.com/28379439/65473180-950df700-de6e-11e9-8188-7e8dc47fdfbd.png)
